### ISSUE TITLE
Unsafe functions

### ISSUE THREAD

**Issue by Disinterpreter** [multitheftauto/mtasa-blue#1972]
**Describe the solution you'd like**

From: https://github.com/multitheftauto/mtasa-blue/pull/1937#issuecomment-752962847 https://github.com/multitheftauto/mtasa-blue/discussions/1782 we decided to allow unsafe functions only with ACL rights.

I'd like to start implementing new functions from ACL fixes.

In the first step we deny all unsafe functions
When we will be sure that these functions blocked by default, we can start to implement ones.

I suppose we should mode current os.* functions under ACL too.

**Additional context**

Put something like this in Default (check me with methods like "file:close")
```xml
<right name="function.require" access="false"></right>
<right name="function.dofile" access="false"></right>
<right name="function.loadfile" access="false"></right>
<right name="function.loadstring" access="false"></right>

<right name="function.os.execute" access="false"></right>
<right name="function.os.exit" access="false"></right>
<right name="function.os.getenv" access="false"></right>
<right name="function.os.remove" access="false"></right>
<right name="function.os.rename" access="false"></right>
<right name="function.os.setlocale" access="false"></right>
<right name="function.os.tmpname" access="false"></right>

<right name="function.io.close" access="false"></right>
<right name="function.io.flush" access="false"></right>
<right name="function.io.input" access="false"></right>
<right name="function.io.lines" access="false"></right>
<right name="function.io.open" access="false"></right>
<right name="function.io.output" access="false"></right>
<right name="function.io.popen" access="false"></right>
<right name="function.io.read" access="false"></right>
<right name="function.io.stderr" access="false"></right>
<right name="function.io.stdin" access="false"></right>
<right name="function.io.stdout" access="false"></right>
<right name="function.io.tmpfile" access="false"></right>
<right name="function.io.type" access="false"></right>
<right name="function.io.write" access="false"></right>
<right name="function.file:close" access="false"></right>
<right name="function.file:flush" access="false"></right>
<right name="function.file:lines" access="false"></right>
<right name="function.file:read" access="false"></right>
<right name="function.file:seek" access="false"></right>
<right name="function.file:setvbuf" access="false"></right>
<right name="function.file:write" access="false"></right>


<right name="function.package.cpath" access="false"></right>
<right name="function.package.loaded" access="false"></right>
<right name="function.package.loaders" access="false"></right>
<right name="function.package.loadlib" access="false"></right>
<right name="function.package.path" access="false"></right>
<right name="function.package.preload" access="false"></right>
<right name="function.package.seeall" access="false"></right>
```

**Comment by xLuxy** [comment]
Lots of functions are already disabled by default - see https://github.com/multitheftauto/mtasa-blue/blob/a0410c68945b6b3d9225c8c90af2bb6b1cda5eaa/Client/mods/deathmatch/logic/lua/CLuaMain.cpp#L75

The following functions are safe to use
### OS

- os.clock
- os.date
- os.time
- os.difftime

(See: https://github.com/multitheftauto/mtasa-blue/blob/a0410c68945b6b3d9225c8c90af2bb6b1cda5eaa/vendor/lua/src/loslib.c#L220)

**Comment by xLive** [comment]
There's no reason to disable time functions by default

**Comment by Disinterpreter** [comment]
> 
> 
> Lots of functions are already disabled by default - see
> 
> https://github.com/multitheftauto/mtasa-blue/blob/a0410c68945b6b3d9225c8c90af2bb6b1cda5eaa/Client/mods/deathmatch/logic/lua/CLuaMain.cpp#L75

Yes, I know. But our conflict was started from agreement about unsafe functions must be restricted by default in ACL. It's more agreement problem than technical.

**Comment by Disinterpreter** [comment]
> 
> 
> There's no reason to disable time functions by default

Thanks, I fixed this.

**Comment by LosFaul** [comment]
I personally would prefer to have unsafe functions loaded via meta.xml,
which would make it far easier to enable it per resource instead of having painfull ACL management.

Besides that ACL would also reduce the performance
and usage of unsafe functions on server side would be advanced topic anyway.

**Comment by patrikjuvonen** [comment]
> I personally would prefer to have unsafe functions loaded via meta.xml,
> which would make it far easier to enable it per resource instead of having painfull ACL management.
> 
> Besides that ACL would also reduce the performance
> and usage of unsafe functions on server side would be advanced topic anyway.

meta.xml based ACL is bad because it doesn't prevent anything (malicious resources, etc.)

**Comment by LosFaul** [comment]
> meta.xml based ACL is bad because it doesn't prevent anything (malicious resources, etc.)

the functions are already unsafe
and if someone doesn't check meta before executing a random resource
then the person is simply not very smart

**Comment by patrikjuvonen** [comment]
> the functions are already unsafe
> and if someone doesn't check meta before executing a random resource
> then the person is simply not very smart

Unfortunately the majority of our playerbase runs unsafe resources. Just because you believe you can do better, doesn't mean they can. ACL has a purpose, which is why it's there.

**Comment by xLive** [comment]
> The following functions are safe to use
> 
> ### OS
> * os.clock
> * os.date
> * os.difftime

I think you forget `os.time`?

**Comment by LosFaul** [comment]
> Unfortunately the majority of our playerbase runs unsafe resources. Just because you believe you can do better, doesn't mean they can. ACL has a purpose, which is why it's there.

with that logic you could also disallow c++ server modules entirely...

**Comment by Disinterpreter** [comment]
> with that logic you could also disallow c++ server modules entirely...

You're totally right lol! But I think it might be in ACL because it is a simpler to implementation way.

Existing modules sdk and blocking dofile/require/os.* is absurdly.

**Comment by patrikjuvonen** [comment]
> with that logic you could also disallow c++ server modules entirely...

Let's not go off-topic with "what's worse".

I'm not going to argue further about the benefits of ACL because it's already implemented in MTA for the sole reason of protecting the server. Resources request rights, and the server grants them. Simple as that.

As discussed here: #1782 and #680, ACL is the way to go.

**Comment by xLuxy** [comment]
Thank you @xLive i might've overseen it :P

**Comment by Disinterpreter** [comment]
Blocked by https://github.com/multitheftauto/mtasa-blue/issues/1983

**Comment by ccw808** [comment]
Even with the ACL there will be a problem with escalating access. e.g. Any resource allowed to use `aclSetRight` can give itself access to `os.open`. So anyone using **runcode** could eventually read/write any file on the hosting server.

**Comment by sbx320** [comment]
I think reimplementing most of those functions in a resource-only context (similar to existing MTA functions) would be better, since it makes using non-MTA Lua scripts a lot easier. Generally I think allowing scripts to load arbitrary binaries is a very bad idea.

### Quick List: 

- `require`, `dofile`, `loadfile` Limit to only work on Lua files
- `loadstring` no changes
- `os.exit` alias for `shutdown`
- `os.remove`, `os.rename` alias to `File.delete`, `File.rename`
- `io.stderr´, `io.stdout`: Could be implemented via an always open file handle which just prints to the console
- Most `io.*` and `file.*` functions can be implemented via `File` as well
- Other functions would also be easy to implement like `io.tmpfile`
- `os.getenv`: Seems fine to me to implement with ACL checks

### The remaining functions would be: 

- `os.setlocale`: Locales are scary, wouldn't implement, not particularly useful either
- `io.stdin´, `file:setvbuf`: Don't really apply to MTA
- `io.popen`, `os.execute`: I'd rather not implement those at all
- `package.*`: Same as with `os.execute`, I'd not implement those

**Comment by Disinterpreter** [comment]
> 
> 
> I think reimplementing most of those functions in a resource-only context (similar to existing MTA functions) would be better, since it makes using non-MTA Lua scripts a lot easier. Generally I think allowing scripts to load arbitrary binaries is a very bad idea.
> ### Quick List:
> 
>     * `require`, `dofile`, `loadfile` Limit to only work on Lua files

But what if I totally sure that I want to use it? 

https://github.com/neoxic/lua-mongo
https://github.com/thibaultcha/lua-cassandra
https://github.com/philanc/luarabbit
https://github.com/DhavalKapil/elasticsearch-lua
https://github.com/cloudflare/raven-lua

(About second, yes, it is a 100% Lua script, but it uses sockets)

I can explain using of each module in the project

**Comment by sbx320** [comment]
I'd say it's up to you to readd that functionality in a server module.

**Comment by Disinterpreter** [comment]
> 
> 
> I'd say it's up to you to readd that functionality in a server module.

I'm so sorry but it is really crap for today. I don't know how the stack of software can change and writing fully compatibility for yet another software in C++ is really long. + thousand minecraft, rage, fivem servers live with more weak rules.

P.S I'm not disagree of any alert, change config etc... But what is the reason that I can't use require to dll it is not good for pro-programmers.

**Comment by jlillis** [comment]
With proper ACL protection, I don't see why `require`, `dofile`, and `loadfile` should be restricted to Lua files only. Allowing modules to be loaded would be a major value-add for this feature. Assuming they are going to be dissallowed by default, a server owner would have to actively "void their warranty" by giving a resource _new_ privileges.

**Comment by Pirulax** [comment]
You dont need ACL for `loadfile` and `dofile`. `loadstring` is available without ACL. What we should do is just write a wrapper around `loadfile` and `dofile` to follow the MTA path stuff (like `:resource/`, etc..), and to only allow signed bytecode to run (or disallow bytecode, whatever, iirc loadstring works with bytecode as well)

**Comment by sbx320** [comment]
> Assuming they are going to be dissallowed by default, a server owner would have to actively "void their warranty" by giving a resource new privileges.

The issue is that some significant number of server owners wild blindly follow random guides to grant scripts full ACL access. Right now the worst thing you could do with a backdoor is to delete some user data or fill the server's hard disk. Allowing resources to load arbitrary binaries goes much further than that. 

Ultimately if a server owner really wants to enable things like require, it's easy to add with a module (just copy the source for require from Lua)

**Comment by prnxdev** [comment]
I started using Haxe which after compilation sometimes needs to use Lua specific libs (like `bit` for example). It's not huge problem because there are functions that are written in MTA API that are bypassing those problems. For example Haxe have native JSON parsing and stringifying functions:

```hx
var json = '{"rating": 5}';
var value = JSON.parse(json); // { rating: 5 }

trace(value.rating); // 5
```

I suggest to add some ability to manage those "unsafe functions". Like disable them by default but give developers ability to unlock it (or if we are using require function only allow require to be used with internal Lua libs and not for files to prevent some security issues).
